driver-core/src/main/com/mongodb/internal/inject/OptionalProvider.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.inject;
+
+import java.util.Optional;
+
+/**
+ * Inspired by <a href="https://jakarta.ee/specifications/dependency-injection/2.0/apidocs/jakarta/inject/provider">
```

Comment:
> That interface seems to be the same as the Supplier interface.
Yes, but this interface expresses the intent more specifically. As a result of the interface being more specific, an automatic injector may implement it in order to fulfill the requested dependencies. In case of a `Suppier`, it would be unclear whether the actual dependency is the `Supplier` itself, or the object it supplies.

I left the comment to show that injecting `Provider` instead of the actual dependency is a standard approach of dealing with dependency cycles between objects. This interface is a [part of Java EE 8](https://javaee.github.io/javaee-spec/javadocs/javax/inject/Provider.html) (the last EE version), and is now a part of the Jakarta EE (a successor of Java EE).

I guess the comment turned out to not be useful, so I will remove it.

Reply:
Done.